thinc/layers/dropout.py

```diff
@@ -18,40 +17,43 @@ def Dropout(rate: float = 0.0) -> Model[InT, InT]:
     return Model("dropout", forward, attrs={"dropout_rate": rate, "is_enabled": True})
 
 
-# We're getting type hell here, I think because of the instance checks?
-# It's sort of painful, because I think this confused the types of other
-# layers that are trying to use dropout.
-# I've relaxed the types for now, but it'd be good to understand what's wrong
-# here.
-def forward(model: Model, X, is_train: bool) -> Tuple[Any, Callable]:
+def forward(
+    model: Model[InT, InT], X: InT, is_train: bool
+) -> Tuple[InT, Callable]:
     rate = model.attrs["dropout_rate"]
     is_enabled = model.attrs["is_enabled"] and is_train
     if rate == 0 or not is_enabled:
-        return X, lambda dY: dY
+        unchanged_return_value, backprop = X, lambda dY: dY
+        return_value = cast(InT, unchanged_return_value)
     elif isinstance(X, Ragged):
-        return _dropout_ragged(model, X, is_train)
+        ragged_return_value, backprop = _dropout_ragged(model, X, is_train)
+        return_value = cast(InT, ragged_return_value)
```

Comment:
Here the problem is that once a variable like `return_value` or `data` has been used once, it retains the type assigned to it in later code even if it has been returned in the meantime. This is why I originally wanted to declare `InT` as the return types of the `_dropout...` methods, because that avoids all such problems in the general method.

The code could still be restructured to return within each `if` clause, but I think a different variable name will have to be used each time if we want to retain the specific return types further below.

Reply:
Ugh, you're right, that's really a limitation of `mypy` though, especially with the variable being used in separate `if` clauses which would all `return`. I'd still like to make this shorter though, so maybe something like this?
```
    if rate == 0 or not is_enabled:
        return X, lambda dY: dY
    elif isinstance(X, Ragged):
        data_r, backprop = _dropout_ragged(model, X, is_train)
        return cast(InT, data_r), backprop
    elif isinstance(X, Padded):
        data_p, backprop = _dropout_padded(model, X, is_train)
        return cast(InT, data_p), backprop
    elif isinstance(X, Sequence):
        data_l, backprop = _dropout_lists(model, X, is_train)
        return cast(InT, data_l), backprop
    else:
        data_a, backprop = _dropout_array(model, cast(ArrayXd, X), is_train)
        return cast(InT, data_a), backprop
```